### DISCUSSION TITLE
Question: non-blocking windows

### DISCUSSION THREAD

**Discussion by schlegelp** [fury-gl/fury#111] [General]
Quick question: I'd like to be able to interact with a scene programmatically from the Python shell (e.g. add/remove actors). Is there a way to show a non-blocking window?

Thanks!
Philipp

**Comment by skoudoro** [comment]
Hi @schlegelp, 

Currently, it is not possible, but we are working on it. it seems we need to use `vtkGenericRenderWindowInteractor`.
Otherwise, if you do not need to show, you can have differents options:
- use `offscreen` option
- play with timer event like [in this tutorial](http://fury.gl/latest/auto_tutorials/viz_timers.html#sphx-glr-auto-tutorials-viz-timers-py)
- record some event and replay them (look at our tests)

hope it helps

**Comment by skoudoro** [comment]
Do you have another idea @Garyfallidis?

**Comment by Garyfallidis** [comment]
Hi Philip (@schlegelp), 
Serge, is correct currently the window is blocking. However, you can have a nonblocking window if you use Qt. Here is an example of how you could do this

```
import sys
import time
import vtk
import threading
import numpy as np
from PyQt5 import QtCore, QtWidgets
from vtk.qt.QVTKRenderWindowInteractor import QVTKRenderWindowInteractor

class MainWindow(QtWidgets.QMainWindow):

    def __init__(self, parent = None):
        QtWidgets.QMainWindow.__init__(self, parent)

        self.frame = QtWidgets.QFrame()

        self.vl = QtWidgets.QVBoxLayout()
        self.vtkWidget = QVTKRenderWindowInteractor(self.frame)
        self.vl.addWidget(self.vtkWidget)

        self.ren = vtk.vtkRenderer()
        self.vtkWidget.GetRenderWindow().AddRenderer(self.ren)
        self.iren = self.vtkWidget.GetRenderWindow().GetInteractor()

        # Create source
        source = vtk.vtkSphereSource()
        source.SetCenter(0, 0, 0)
        source.SetRadius(5.0)

        # Create a mapper
        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputConnection(source.GetOutputPort())

        # Create an actor
        actor = vtk.vtkActor()
        actor.SetMapper(mapper)

        self.ren.AddActor(actor)

        self.ren.ResetCamera()

        self.frame.setLayout(self.vl)
        self.setCentralWidget(self.frame)

        self.show()
        self.iren.Initialize()

def calc_square(numbers, window):
    print("calculate square numbers")
    for n in numbers:
        time.sleep(1)
        print('square:', n*n)
        window.ren.GetActiveCamera().Azimuth(10)
        window.vtkWidget.Render()

if __name__ == "__main__":

    app = QtWidgets.QApplication(sys.argv)

    window = MainWindow()

    arr = np.random.rand(10)
    t1 = threading.Thread(target=calc_square, args=(arr, window))

    t1.start()

    sys.exit(app.exec_())

    t1.join()
```

We would love some help in allowing show and ShowManager to switch different backends including Qt. Let us know if you can help with a Qt backend.

**Comment by Garyfallidis** [comment]
In the meantime, we are also working on trying to figure out how to make the default window non-blocking. Not trivial.... but there is progress.

**Comment by schlegelp** [comment]
Thanks for the responses. I'll let you know if I come up with something useful!
